src/psbasemap.1
.TH PSBASEMAP 1 "Feb 27 2014" "GMT 4.5.13 (SVN)" "Generic Mapping Tools"
.SH NAME
psbasemap \- To plot \fIPostScript\fP basemaps
.SH SYNOPSIS
\fBpsbasemap\fP \fB\-B\fP[\fBp\fP|\fBs\fP]\fIparameters\fP \fB\-J\fP\fIparameters\fP \fB\-R\fP\fIwest\fP/\fIeast\fP/\fIsouth\fP/\fInorth\fP[/\fIzmin\fP/\fIzmax\fP][\fBr\fP] 
[ \fB\-E\fP\fIazim\fP/\fIelev\fP[\fB+w\fP\fIlon\fP/\fIlat\fP[/\fIz\fP]][\fB+v\fP\fIx0\fP/\fIy0\fP] ] [ \fB\-G\fP\fIfill\fP ] [ \fB\-Jz\fP|\fBZ\fP\fIparameters\fP ] [ \fB\-K\fP ] 
[ \fB\-L\fP[\fBf\fP][\fBx\fP]\fIlon0\fP/\fIlat0\fP[/\fIslon\fP]/\fIslat\fP/\fIlength\fP[\fBm\fP|\fBn\fP|\fBk\fP][\fB+l\fP\fIlabel\fP][\fB+j\fP\fIjust\fP][\fB+p\fP\fIpen\fP][\fB+f\fP\fIfill\fP][\fB+u\fP] ] ] [ \fB\-O\fP ] [ \fB\-P\fP ] [ \fB\-U\fP[\fIjust\fP/\fIdx\fP/\fIdy\fP/][\fBc\fP|\fIlabel\fP] ] 
[ \fB\-T\fP[\fBf\fP|\fBm\fP][\fBx\fP]\fIlon0\fP/\fIlat0\fP/\fIsize\fP[/\fIinfo\fP][\fB:\fPw,e,s,n\fB:\fP][\fB+\fP\fIgint\fP[/\fImint\fP]] ] [ \fB\-V\fP ] [ \fB\-X\fP[\fBa\fP|\fBc\fP|\fBr\fP][\fIx-shift\fP[\fBu\fP]] ] [ \fB\-Y\fP[\fBa\fP|\fBc\fP|\fBr\fP][\fIy-shift\fP[\fBu\fP]] ] [ \fB\-Z\fP\fIzlevel\fP ] [ \fB\-c\fP\fIcopies\fP ]
.SH DESCRIPTION
\fBpsbasemap\fP creates \fIPostScript\fP code that will produce a basemap.  Several map projections
are available, and the user may specify separate tickmark intervals for boundary annotation,
ticking, and [optionally] gridlines.  A simple map scale or directional rose may also be plotted.
.TP
\fB\-B\fP
Sets map boundary annotation and tickmark intervals. The format of \fItickinfo\fP is
[\fBp\fP|\fBs\fP]\fIxinfo\fP[/\fIyinfo\fP[/\fIzinfo\fP]][\fB:.\fP"Title"\fB:\fP][\fBW\fP|\fBw\fP][\fBE\fP|\fBe\fP][\fBS\fP|\fBs\fP][\fBN\fP|\fBn\fP][\fBZ\fP|\fBz\fP[\fB+\fP]].
The leading \fBp\fP [Default] or \fBs\fP selects the primary or secondary annotation information.
Each of the \fI?info\fP segments are textstrings of the form
\fIinfo\fP[\fB:\fP"Axis label"\fB:\fP][\fB:=\fP"prefix"\fB:\fP][\fB:,\fP"unit label"\fB:\fP].
The \fIinfo\fP string is made up of one or more concatenated substrings of the form
[\fBa\fP|\fBf\fP|\fBg\fP]\fIstride\fP[+-\fIphase\fP][\fIunit\fP].
The leading \fBa\fP is used to specify the annotation and major tick spacing [Default], \fBf\fP for minor tick spacing, and
\fBg\fP for gridline spacing.
\fIstride\fP is the desired stride interval.
The optional \fIphase\fP shifts the annotation interval by that amount (positive or negative).
The optional \fIunit\fP indicates the unit of the \fIstride\fP and can be any of
\fBY\fP (year, plot with 4 digits), \fBy\fP (year, plot with 2 digits), \fBO\fP (month, plot using \fBPLOT_DATE_FORMAT\fP),
\fBo\fP (month, plot with 2 digits), \fBU\fP (ISO week, plot using \fBPLOT_DATE_FORMAT\fP), \fBu\fP (ISO week, plot using 2 digits),
\fBr\fP (Gregorian week, 7-day stride from start of week \fBTIME_WEEK_START\fP), \fBK\fP (ISO weekday, plot name of day),
\fBD\fP (date, plot using \fBPLOT_DATE_FORMAT\fP), \fBd\fP (day, plot day of month 0-31 or year 1-366, via \fBPLOT_DATE_FORMAT\fP),
\fBR\fP (day, same as \fBd\fP, aligned with \fBTIME_WEEK_START\fP), \fBH\fP (hour, plot using \fBPLOT_CLOCK_FORMAT\fP),
\fBh\fP (hour, plot with 2 digits), \fBM\fP (minute, plot using \fBPLOT_CLOCK_FORMAT\fP), \fBm\fP (minute, plot with 2 digits), 
\fBC\fP (second, plot using \fBPLOT_CLOCK_FORMAT\fP), \fBc\fP (second, plot with 2 digits).  Note for geographic axes
\fBm\fP and \fBc\fP instead mean arc minutes and arc seconds.  All entities that are language-specific are under control
by \fBTIME_LANGUAGE\fP.  To specify separate x and y ticks, separate the substrings that apply to the x and y
axes with a slash [/] (If a 3-D basemap is selected with \fB\-E\fP and \fB\-Jz\fP, a third substring pertaining
to the vertical axis may be appended.)  For linear/log/power projections (\fB\-Jx\fP|\fBX\fP): Labels
for each axis can be added by surrounding them with colons (\fB:\fP).  If the first character in the
label is a period, then the label is used as plot title; if it is a comma (\fB,\fP) then the label is
appended to each annotation; if it is an equal sign (\fB=\fP) the the prefix is prepended to each annotation (start label/prefix
with - to avoid space between annotation and item); else it is the axis label.
If the label consists of more than one word, enclose the entire label in double quotes (e.g.,
\fB:\fP"my label"\fB:\fP).
If you need to use a colon (:) as part of your label you must specify it using its octal code (\\072).
.br
By default, all 4 boundaries are plotted (referred to as \fBW, E, S, N\fP).  To change the default,
append the code for only those axes you want (e.g., \fBWS\fP for standard lower-left x- and y-axis system).
Upper case (e.g., \fBW\fP) means draw axis/tickmarks AND annotate it, whereas lower case
(e.g., \fBw\fP) will only draw axis/tickmarks.  (If a 3-D basemap is selected with \fB\-E\fP and
\fB\-Jz\fP, append \fBZ\fP or \fBz\fP to control the appearance
of the vertical axis.  Append \fB+\fP to draw the outline of the cube defined by \fB\-R\fP.  Note
that for 3-D views the title, if given, will be suppressed.)
.br
For non-geographical projections:
Give negative scale (in \fB\-Jx\fP) or axis length (in \fB\-JX\fP) to change the direction
of increasing coordinates (i.e., to make the y-axis positive down).  For log10 axes:  Annotations
can be specified in one of three ways:  (1) \fIstride\fP can be 1, 2, 3, or -\fIn\fP.  Annotations will then
occur at 1, 1\-2\-5, or 1\-2\-3\-4\-...\-9, respectively; for -\fIn\fP we annotate every \fIn\fP't magnitude.
This option can also be used for
the frame and grid intervals.  (2) An \fBl\fP is appended to the \fItickinfo\fP string.
Then, log10 of the tick value is plotted at every integer log10 value.  (3) A \fBp\fP is appended
to the \fItickinfo\fP string.  Then, annotations appear as 10 raised to log10 of the tick value.
For power axes: Annotations can be specified in one of two ways:  (1) \fIstride\fP sets the
regular annotation interval.  (2) A \fBp\fP is appended to the \fItickinfo\fP string.  Then,
the annotation interval is expected to be in transformed units, but the annotation value
will be plotted as untransformed units.  E.g., if \fIstride\fP = 1 and \fIpower\fP = 0.5 (i.e., sqrt),
then equidistant annotations labeled 1\-4\-9... will appear.
.br
These GMT parameters can affect the appearance of the map boundary: \fBANNOT_MIN_ANGLE\fP, \fBANNOT_MIN_SPACING\fP,
\fBANNOT_FONT_PRIMARY\fP, \fBANNOT_FONT_SECONDARY\fP, \fBANNOT_FONT_SIZE_PRIMARY\fP, \fBANNOT_FONT_SIZE_SECONDARY\fP,
\fBANNOT_OFFSET_PRIMARY\fP, \fBANNOT_OFFSET_SECONDARY\fP, \fBBASEMAP_AXES\fP, \fBBASEMAP_FRAME_RGB\fP, \fBBASEMAP_TYPE\fP, \fBPLOT_DEGREE_FORMAT\fP,
\fBFRAME_PEN\fP, \fBFRAME_WIDTH\fP, \fBGRID_CROSS_SIZE_PRIMARY\fP, \fBGRID_PEN_PRIMARY\fP, \fBGRID_CROSS_SIZE_SECONDARY\fP,
\fBGRID_PEN_SECONDARY\fP, \fBHEADER_FONT\fP, \fBHEADER_FONT_SIZE\fP, \fBLABEL_FONT\fP, \fBLABEL_FONT_SIZE\fP, \fBLINE_STEP\fP,
\fBOBLIQUE_ANNOTATION\fP, \fBPLOT_CLOCK_FORMAT\fP, \fBPLOT_DATE_FORMAT\fP, \fBTIME_FORMAT_PRIMARY\fP, \fBTIME_FORMAT_SECONDARY\fP,
\fBTIME_LANGUAGE\fP, \fBTIME_WEEK_START\fP, \fBTICK_LENGTH\fP, \fBTICK_PEN\fP, and \fBY_AXIS_TYPE\fP; see the
\fBgmtdefaults\fP man page for details.
.TP
\fB\-J\fP
Selects the map projection. The following character determines the projection. If the
character is upper case then the argument(s) supplied as scale(s) is interpreted to be
the map width (or axis lengths), else the scale argument(s) is the map scale (see its
definition for each projection). UNIT is cm, inch, or m, depending on the \fBMEASURE_UNIT\fP
setting in \.gmtdefaults4, but this can be overridden on the command line by appending
\fBc\fP, \fBi\fP, or \fBm\fP to the \fIscale\fP or \fIwidth\fP values.  Append \fBh\fP, \fB+\fP, or \fB-\fP
to the given \fIwidth\fP if you instead want to set map height, the maximum dimension, or
the minimum dimension, respectively [Default is \fBw\fP for width].
.br
In case the central meridian is an optional parameter and it is being omitted, then the
center of the longitude range given by the \fB\-R\fP option is used. The default standard parallel
is the equator.
.br
The ellipsoid used in the map projections is user-definable by editing the \.gmtdefaults4 file
in your home directory. 73 commonly used ellipsoids and spheroids are currently
supported, and users may also specify their own custum ellipsoid parameters [Default is WGS-84].
Several GMT parameters can affect the projection: \fBELLIPSOID\fP, \fBINTERPOLANT\fP,
\fBMAP_SCALE_FACTOR\fP, and \fBMEASURE_UNIT\fP; see the \fBgmtdefaults\fP man page for details.
.br
Choose one of the following projections (The \fBE\fP or \fBC\fP after projection names
stands for Equal-Area and Conformal, respectively):
.RS
.PP
\fBCYLINDRICAL PROJECTIONS:\fP
.TP
\fB\-Jc\fP\fIlon0/lat0/scale\fP or \fB\-JC\fP\fIlon0/lat0/width\fP (Cassini).
Give projection center \fIlon0/lat0\fP and \fIscale\fP (\fB1:\fP\fIxxxx\fP or UNIT/degree).
.TP
\fB\-Jcyl_stere\fP/[\fIlon0/\fP[\fIlat0/\fP]]\fIscale\fP or \fB\-JCyl_stere\fP/[\fIlon0/\fP[\fIlat0/\fP]]\fIwidth\fP (Cylindrical Stereographic).
Give central meridian \fIlon0\fP (optional), standard parallel \fIlat0\fP (optional), and \fIscale\fP along parallel (\fB1:\fP\fIxxxx\fP or UNIT/degree).
The standard parallel is typically one of these (but can be any value):
.RS
.RS
66.159467 - Miller's modified Gall
.br
55 - Kamenetskiy's First
.br
45 - Gall's Stereographic
.br
30 - Bolshoi Sovietskii Atlas Mira or Kamenetskiy's Second
.br
0 - Braun's Cylindrical
.RE
.RE
.TP
\fB\-Jj\fP[\fIlon0/\fP]\fIscale\fP or \fB\-JJ\fP[\fIlon0/\fP]\fIwidth\fP (Miller Cylindrical Projection).
Give the central meridian \fIlon0\fP (optional) and \fIscale\fP (\fB1:\fP\fIxxxx\fP or UNIT/degree).
.TP
\fB\-Jm\fP[\fIlon0/\fP[\fIlat0/\fP]]\fIscale\fP or \fB\-JM\fP[\fIlon0/\fP[\fIlat0/\fP]]\fIwidth\fP
Give central meridian \fIlon0\fP (optional), standard parallel \fIlat0\fP (optional), and \fIscale\fP along parallel (\fB1:\fP\fIxxxx\fP or UNIT/degree).
.TP
\fB\-Jo\fP\fIparameters\fP (Oblique Mercator \fB[C]\fP).
Typically used with \fB\-R\fP<...>\fBr\fP, otherwise region is in oblique coordinates. Specify one of:
.RS
.TP
\fB\-Jo\fP[\fBa\fP]\fIlon0/lat0/azimuth/scale\fP or \fB\-JO\fP[\fBa\fP]\fIlon0/lat0/azimuth/width\fP
Set projection center \fIlon0/lat0\fP, \fIazimuth\fP of oblique equator, and \fIscale\fP.
.TP
\fB\-Jo\fP[\fBb\fP]\fIlon0/lat0/lon1/lat1/scale\fP or \fB\-JO\fP[\fBb\fP]\fIlon0/lat0/lon1/lat1/scale\fP
Set projection center \fIlon0/lat0\fP, another point on the oblique equator \fIlon1/lat1\fP, and \fIscale\fP.
.TP
\fB\-Joc\fP\fIlon0/lat0/lonp/latp/scale\fP or \fB\-JOc\fP\fIlon0/lat0/lonp/latp/scale\fP
Set projection center \fIlon0/lat0\fP, pole of oblique projection \fIlonp/latp\fP, and \fIscale\fP.
.PP
Give \fIscale\fP along oblique equator (\fB1:\fP\fIxxxx\fP or UNIT/degree).
.RE
.TP
\fB\-Jq\fP[\fIlon0/\fP[\fIlat0/\fP]]\fIscale\fP or \fB\-JQ\fP[\fIlon0/\fP[\fIlat0/\fP]]\fIwidth\fP (Cylindrical Equidistant).
Give the central meridian \fIlon0\fP (optional), standard parallel \fIlat0\fP (optional), and \fIscale\fP (\fB1:\fP\fIxxxx\fP or UNIT/degree).
The standard parallel is typically one of these (but can be any value):
.RS
.RS
61.7 - Grafarend and Niermann, minimum linear distortion
.br
50.5 - Ronald Miller Equirectangular
.br
43.5 - Ronald Miller, minimum continental distortion
.br
42 - Grafarend and Niermann
.br
37.5 - Ronald Miller, minimum overall distortion
.br
0 - Plate Carree, Simple Cylindrical, Plain/Plane Chart
.br
.RE
.RE
.TP
\fB\-Jt\fP\fIlon0/\fP[\fIlat0/\fP]\fIscale\fP or \fB\-JT\fP\fIlon0/\fP[\fIlat0/\fP]\fIwidth\fP
Give the central meridian \fIlon0\fP, central parallel \fIlat0\fP (optional), and \fIscale\fP (\fB1:\fP\fIxxxx\fP or UNIT/degree).
.TP
\fB\-Ju\fP\fIzone/scale\fP or \fB\-JU\fP\fIzone/width\fP (UTM - Universal Transverse Mercator \fB[C]\fP).
Give the UTM zone (A,B,1-60[C-X],Y,Z)) and \fIscale\fP (\fB1:\fP\fIxxxx\fP or UNIT/degree).
.br
Zones: If C-X not given, prepend - or + to enforce southern or northern hemisphere conventions [northern if south > 0].
.TP
\fB\-Jy\fP[\fIlon0/\fP[\fIlat0/\fP]]\fIscale\fP or \fB\-JY\fP[\fIlon0/\fP[\fIlat0/\fP]]\fIwidth\fP (Cylindrical Equal-Area \fB[E]\fP).
Give the central meridian \fIlon0\fP (optional), standard parallel \fIlat0\fP (optional), and \fIscale\fP (\fB1:\fP\fIxxxx\fP or UNIT/degree).
The standard parallel is typically one of these (but can be any value):
.RS
.RS
50 - Balthasart
.br
45 - Gall-Peters
.br
37.0666 - Caster
.br
37.4 - Trystan Edwards
.br
37.5 - Hobo-Dyer
.br
30 - Behrman
.br
0 - Lambert (default)
.RE
.RE
.PP
\fBCONIC PROJECTIONS:\fP
.TP
\fB\-Jb\fP\fIlon0/lat0/lat1/lat2/scale\fP or \fB\-JB\fP\fIlon0/lat0/lat1/lat2/width\fP (Albers \fB[E]\fP).
Give projection center \fIlon0/lat0\fP, two standard parallels \fIlat1/lat2\fP, and \fIscale\fP (\fB1:\fP\fIxxxx\fP or UNIT/degree).
.TP

ChangeLog
-----------------------------------------------------------------------------------------
	$Id: ChangeLog 12925 2014-02-20 22:02:24Z pwessel $
-----------------------------------------------------------------------------------------

-----------------------------------------------------------------------------------------
201x-yy-zz	pwessel
	GMT 5.1.1 finalized and released.
-----------------------------------------------------------------------------------------
2014-02-20	pwessel
	* grdcontour.c:		The -Zscale/offset option was applied incorrectly due to a
				change in the underlying scaling function.
	* grdvolume.c:		---"---
2014-02-10	pwessel
	*gmt2kml.c:		Make sure the default style IDs are unique for each process.

2014-02-03	pwessel
	* pslib.c:		Auto-widen paragraph width to at least fit widest word.

2014-01-26	pwessel
	* gmt_grdio.c:		Building the grid command string could exceed its length if
				the argument became exactly the max length of the array; then
				a final extra character \0 would be added and exceed array length.
				Thanks to Joachim Saul for pointing it out.

2014-01-22	pwessel
	* pscoast.c:		The processing of -I -N levels and pens could get out of sync
				so that the wrong pen was used for the specified feature.
2014-01-16	remko
	* pslib.c:		Take superscripts and subscripts into account when determining
				dimensions of text box.

2014-01-10	pwessel
	* grdmath.c:		We consider the change in SDIST, PDIST to return degrees instead
				of km to be a bug as it makes it difficult to handle geodesics
				for which degrees make no sense.  Thus we have consolidated all
				the ?DIST operators to give km if geographic data are involved.
				Example 23 has been adjusted accordingly, as has several test
				scripts.  Please check your own scripts for SDIST.

2014-01-09	pwessel
	* gmt_vector.c:		Add protection against introduced longitude-jumps produced
				by resampling in GMT_fix_up_path.
2014-01-09	remko
	* gmt_vector.c:		Add protection against introduced longitude-jumps produced
				by resampling in GMT_fix_up_path.
	* grdmath.c:		LDIST and PDIST will again produce km when using geodesics.

2014-01-01	pwessel
	* x2sys/x2sys.c:	Mishandled the assignment of segment number for each record.
	* gmt_support.c:	Allow GMT_crossover to deal with global wrapped data.

2013-12-31	pwessel
	* gmt2kml.c:		Now using GMT_Put_Record instead of old stdout printing.
				This addresses issue #473.

2013-12-10	pwessel
	* grdvector.c:		Vector direction did not adjust when negative scales were
				used in -JX or -Jx.
	* psxy[z].c		---"---

2013-12-08	remko
	* psimage.c:		Now properly plots transparent images and makes -Gt work.

2013-11-26	pwessel
	* gmt_io.c:		GMT would crash on a blank line with no newline at end at all.
	* pstext.c:		pstext did not process GMT4-style input properly.

2013-11-26	jluis
	* potential/gravfft.c:	Let -E options default to 3.
				Let users select the -N+l (leave trend alone). That is, do not overwrite with default mode.
				Let the default detrend be remove MID point.
				Explicitly set -N+a in gravfft command to avoid failures due to coming changes in gravfft.
2013-11-25	pwessel
	* grdlandmask.c:	Failed to fill partial bins with no coastlines when the region
				crossed Greenwich [issue #451].

2013-11-21	pwessel
	* pscontour.c:		Let pscontour output contours to file [issue #448].
	* pscoast.c:		Prevent users from not giving proper projection [issue #449].

2013-11-19	jluis
	* mgd77/mgd77magref.c:	Needed to reset the t_col variable when constant altitude [issue #446].

2013-11-19	pwessel
	* gmt_sph.c:		Fix invalid read in ssrfpack_grid due to par array being of length 2 rather than 3.

2013-11-18	pwessel
	* gmt_grdio.c:		Place the old auto-detect of 4-byte binary float vs int under compatibility mode for backwardness.

2013-11-16	pwessel
	* grdvector.c:		Address shortcomings of geovectors when wrapped and clipped.

2013-11-15	pwessel
	* geovectors.c:		Fix geovector when shrinking is selected.

2013-11-13	pwessel
	* gmt_api.c:		Fix API issues pointed out by Ian.

2013-11-13	fwobbe
	* GMT.app:		Inherit user shell in GMT.app.

2013-11-12	pwessel
	* gmt_init.c:		grdmath could corrupt the gmt.history file [issue # 438].
 	
2013-11-10	pwessel
	* gmt_support.c:	Fix symbol-size issue in psxy -Sc [issue #435].
	
2013-11-09	pwessel
	* grdvector.c:		Did not convert direction to azimuths for geo vectors.

2013-11-08	pwessel
	* grdvector.c:		Fix grdvector -I problem [issue #434].
	* ps2raster.c:		Under Mavericks, strcpy would exit if the to and from pointers
-----------------------------------------------------------------------------------------
2013-11-05	pwessel
	GMT 5.1.0 finalized and released.
-----------------------------------------------------------------------------------------

2013-11-01	remko
	* example_10.sh:	Changed input to number of languages; multi-color columns

2013-10-31	pwessel
	* gmt_dcw.c:		Added option to specify code as =<continent> which will
				automatically be expanded to all the countries in that
				continent.  E.g., -F=AF+gred will paint Africa read.
				Fixed memory leaks.

2013-10-23	pwessel
	* gmtinfo.c:		The new name of minmax.c, the latter will be honored
				when GMT compatibility mode is 4.
	* grdmath.c:		Fixed porting bugs for CURV and D2DY2; added test ops.sh

2013-10-22	pwessel
	* minmax.c:		Added -D to adjust region found with -I
	* makecpt.c:		Better reporting of gaps or overlaps in CPTs

2013-10-05	pwessel
	* greenspline.c:	Needed -f in order to select -f0T input, plus it made assumptions
				about lon,lat despite not being selected.

2013-10-04	pwessel
	* gmt.history:		This replaces the old .gmtcommands file name.

2013-09-28	remko
	* gmt_init.c:		Moved origin of JD by one day (24 Nov instead of 25 Nov).
	* ConfigUserTemplate.cmake:	Added option GMT_DOCS_DEPEND_ON_GMT

2013-09-27	pwessel
	* share/cpt:		Added 10 new CPTs, mostly for bathymetry and topography.

2013-09-14	pwessel

src/psbasemap.c
/*--------------------------------------------------------------------
 *	$Id: psbasemap.c 12822 2014-01-31 23:39:56Z remko $
 *
 *	Copyright (c) 1991-2014 by P. Wessel, W. H. F. Smith, R. Scharroo, J. Luis and F. Wobbe
 *	See LICENSE.TXT file for copying and redistribution conditions.
 *
 *	This program is free software; you can redistribute it and/or modify
 *	it under the terms of the GNU Lesser General Public License as published by
 *	the Free Software Foundation; version 3 or any later version.
 *
 *	This program is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU Lesser General Public License for more details.
 *
 *	Contact info: gmt.soest.hawaii.edu
 *--------------------------------------------------------------------*/
/*
 * API functions to support the psbasemap application.
 *
 * Author:	Paul Wessel
 * Date:	1-JAN-2010
 * Version:	5 API
 *
 * Brief synopsis: psbasemap plots a basemap for the given area using
 * the specified map projection.
 */

#define THIS_MODULE_NAME	"psbasemap"
#define THIS_MODULE_LIB		"core"
#define THIS_MODULE_PURPOSE	"Plot PostScript base maps"

#include "gmt_dev.h"

#define GMT_PROG_OPTIONS "->BJKOPRUVXYcfptxy" GMT_OPT("EZ")

/* Control structure for psbasemap */

struct PSBASEMAP_CTRL {
	struct D {	/* -D */
		bool active;
		struct GMT_MAP_INSERT item;
	} D;
	struct L {	/* -L */
		bool active;
		struct GMT_MAP_SCALE item;
	} L;
	struct T {	/* -T */
		bool active;
		struct GMT_MAP_ROSE item;
	} T;
};

void *New_psbasemap_Ctrl (struct GMT_CTRL *GMT) {	/* Allocate and initialize a new control structure */
	struct PSBASEMAP_CTRL *C;

	C = GMT_memory (GMT, NULL, 1, struct PSBASEMAP_CTRL);

	/* Initialize values whose defaults are not 0/false/NULL */
	GMT_memset (&C->D.item, 1, struct GMT_MAP_INSERT);
	GMT_memset (&C->L.item, 1, struct GMT_MAP_SCALE);
	GMT_memset (&C->T.item, 1, struct GMT_MAP_ROSE);

	return (C);
}

void Free_psbasemap_Ctrl (struct GMT_CTRL *GMT, struct PSBASEMAP_CTRL *C) {	/* Deallocate control structure */
	if (!C) return;
	GMT_free (GMT, C);
}

int GMT_psbasemap_usage (struct GMTAPI_CTRL *API, int level)
{
	/* This displays the psbasemap synopsis and optionally full usage information */

	GMT_show_name_and_purpose (API, THIS_MODULE_LIB, THIS_MODULE_NAME, THIS_MODULE_PURPOSE);
	if (level == GMT_MODULE_PURPOSE) return (GMT_NOERROR);
	GMT_Message (API, GMT_TIME_NONE, "usage: psbasemap %s %s %s\n", GMT_B_OPT, GMT_J_OPT, GMT_Rgeoz_OPT);
	GMT_Message (API, GMT_TIME_NONE, "\t[-D%s]\n\t[%s] [-K]\n", GMT_INSERT, GMT_Jz_OPT);
	GMT_Message (API, GMT_TIME_NONE, "\t[-L%s]\n", GMT_SCALE);
	GMT_Message (API, GMT_TIME_NONE, "\t[-O] [-P] [-T%s]\n", GMT_TROSE);
	GMT_Message (API, GMT_TIME_NONE, "\t[%s] [%s] [%s]\n\t[%s] [%s] [%s]\n\t[%s] [%s]\n\n", GMT_U_OPT, GMT_V_OPT,
		GMT_X_OPT, GMT_Y_OPT, GMT_c_OPT, GMT_f_OPT, GMT_p_OPT, GMT_t_OPT);

	if (level == GMT_SYNOPSIS) return (EXIT_FAILURE);

	GMT_Option (API, "B,JZ,R");
	GMT_Message (API, GMT_TIME_NONE, "\n\tOPTIONS:\n");
	GMT_mapinsert_syntax (API->GMT, 'D', "Draw a simple map insert box as specified below:");
	GMT_Option (API, "K");
	GMT_mapscale_syntax (API->GMT, 'L', "Draw a simple map scale centered on <lon0>/<lat0>.");
	GMT_Option (API, "O,P");
	GMT_maprose_syntax (API->GMT, 'T', "Draw a north-pointing map rose centered on <lon0>/<lat0>.");
	GMT_Option (API, "U,V,X,c,f,p,t,.");

	return (EXIT_FAILURE);
}

int GMT_psbasemap_parse (struct GMT_CTRL *GMT, struct PSBASEMAP_CTRL *Ctrl, struct GMT_OPTION *options)
{
	/* This parses the options provided to psbasemap and sets parameters in Ctrl.
	 * Note Ctrl has already been initialized and non-zero default values set.
	 * Any GMT common options will override values set previously by other commands.
	 * It also replaces any file names specified as input or output with the data ID
	 * returned when registering these sources/destinations with the API.
	 */

	unsigned int n_errors = 0;
	struct GMT_OPTION *opt = NULL;
	struct GMTAPI_CTRL *API = GMT->parent;

	for (opt = options; opt; opt = opt->next) {	/* Process all the options given */

		switch (opt->option) {

			/* Processes program-specific parameters */

			case 'D':	/* Draw map insert */
				Ctrl->D.active = true;
				n_errors += GMT_getinsert (GMT, 'D', opt->arg, &Ctrl->D.item);
				break;
			case 'G':	/* Set canvas color */
				if (GMT_compat_check (GMT, 4)) {
					GMT_Report (API, GMT_MSG_COMPAT, "Warning: Option -G is deprecated; -B...+g%s was set instead, use this in the future.\n", opt->arg);
					GMT->current.map.frame.paint = true;
					if (GMT_getfill (GMT, opt->arg, &GMT->current.map.frame.fill)) {
						GMT_fill_syntax (GMT, 'G', " ");
						n_errors++;
					}
				}
				else
					n_errors += GMT_default_error (GMT, opt->option);
				break;
			case 'L':	/* Draw map scale */
				Ctrl->L.active = true;
				n_errors += GMT_getscale (GMT, 'L', opt->arg, &Ctrl->L.item);
				break;
			case 'T':	/* Draw map rose */
				Ctrl->T.active = true;
				n_errors += GMT_getrose (GMT, 'T', opt->arg, &Ctrl->T.item);
				break;

			default:	/* Report bad options */
				n_errors += GMT_default_error (GMT, opt->option);
				break;
		}
	}

	n_errors += GMT_check_condition (GMT, !GMT->common.J.active, "Syntax error: Must specify a map projection with the -J option\n");
	n_errors += GMT_check_condition (GMT, !GMT->common.R.active, "Syntax error: Must specify -R option\n");
	n_errors += GMT_check_condition (GMT, !(GMT->current.map.frame.init || Ctrl->D.active || Ctrl->L.active || Ctrl->T.active), "Syntax error: Must specify at least one of -B, -D, -L, -T\n");
	n_errors += GMT_check_condition (GMT, Ctrl->L.active && !GMT_is_geographic (GMT, GMT_IN), "Syntax error: -L applies to geographical data only\n");

	return (n_errors ? GMT_PARSE_ERROR : GMT_OK);
}

#define bailout(code) {GMT_Free_Options (mode); return (code);}
#define Return(code) {Free_psbasemap_Ctrl (GMT, Ctrl); GMT_end_module (GMT, GMT_cpy); bailout(code);}

int GMT_psbasemap (void *V_API, int mode, void *args)
{	/* High-level function that implements the psbasemap task */
	int error;
	
	struct PSBASEMAP_CTRL *Ctrl = NULL;	/* Control structure specific to program */
	struct GMT_CTRL *GMT = NULL, *GMT_cpy = NULL;		/* General GMT interal parameters */
	struct GMT_OPTION *options = NULL;
	struct GMTAPI_CTRL *API = GMT_get_API_ptr (V_API);	/* Cast from void to GMTAPI_CTRL pointer */

	/*----------------------- Standard module initialization and parsing ----------------------*/

	if (API == NULL) return (GMT_NOT_A_SESSION);
	if (mode == GMT_MODULE_PURPOSE) return (GMT_psbasemap_usage (API, GMT_MODULE_PURPOSE));	/* Return the purpose of program */
	options = GMT_Create_Options (API, mode, args);	if (API->error) return (API->error);	/* Set or get option list */

	if (!options || options->option == GMT_OPT_USAGE) bailout (GMT_psbasemap_usage (API, GMT_USAGE));	/* Return the usage message */
	if (options->option == GMT_OPT_SYNOPSIS) bailout (GMT_psbasemap_usage (API, GMT_SYNOPSIS));	/* Return the synopsis */

	/* Parse the command-line arguments; return if errors are encountered */

	GMT = GMT_begin_module (API, THIS_MODULE_LIB, THIS_MODULE_NAME, &GMT_cpy); /* Save current state */
	if (GMT_Parse_Common (API, GMT_PROG_OPTIONS, options)) Return (API->error);
	Ctrl = New_psbasemap_Ctrl (GMT);	/* Allocate and initialize a new control structure */
	if ((error = GMT_psbasemap_parse (GMT, Ctrl, options))) Return (error);

	/*---------------------------- This is the psbasemap main code ----------------------------*/

	/* Ready to make the plot */

	GMT_Report (API, GMT_MSG_VERBOSE, "Constructing the basemap\n");

	if (GMT_err_pass (GMT, GMT_map_setup (GMT, GMT->common.R.wesn), "")) Return (GMT_RUNTIME_ERROR);

	GMT_plotinit (GMT, options);

	GMT_plane_perspective (GMT, GMT->current.proj.z_project.view_plane, GMT->current.proj.z_level);

	GMT_plotcanvas (GMT);	/* Fill canvas if requested */

	GMT_map_basemap (GMT);	/* Plot base map */

	if (Ctrl->D.active) GMT_draw_map_insert (GMT, &Ctrl->D.item);
	if (Ctrl->L.active) GMT_draw_map_scale (GMT, &Ctrl->L.item);
	if (Ctrl->T.active) GMT_draw_map_rose (GMT, &Ctrl->T.item);

	GMT_plane_perspective (GMT, -1, 0.0);

	GMT_plotend (GMT);

	Return (GMT_OK);
}
